### DISCUSSION TITLE
Hi curl team, curl verion 8.9.1 has added new feature: "graceful" connection shutdown, that will call SOCKETFUNCTION with internal curl instance, do other calls of SOCKETFUNCTION for IO(send/receive/remove) changed?

### DISCUSSION THREAD

**Discussion by TheBetterSolution** [curl/curl#14956] [Q&A]
Hi curl team,

You have answered the question:
https://github.com/curl/curl/issues/14767

> - while your transfer (easy handle) is ongoing, the socket/timer callbacks get call with that easy handle and you can retrieve your private data.
> 
> - when the easy handle is done, the connection is placed into the connection pool and kept there for reuse.
> 
> - at some point in time, the connection will be discarded, e.g. shut down and closed. During this shutdown, the socket/timer callbacks will be invoked with an internal easy handle that does not have any PRIVATE data set.
> 
> Since the original easy handle is done and probably has been cleaned up, curl cannot make use of it or its settings any longer. It would be unsafe to return the PRIVATE data somehow, since it is unclear if it disappeared together with its easy handle or not.

Thanks a lot.

I compared the 8.9.1 with older version, I found curl 8.9.1 used the internal instance to call the callback:
https://github.com/curl/curl/blob/8ca60308398e42d77d140ebe5c3331c6e086d355/lib/conncache.c#L771

 if(cpool_update_shutdown_ev(data->multi, cpool->idata, conn)) { 

https://github.com/curl/curl/blob/8ca60308398e42d77d140ebe5c3331c6e086d355/lib/conncache.c#L1114

 if(done || cpool_update_shutdown_ev(multi, data, conn)) { 

But 8.7.1 doesn't call the callback at the same context, so, version 8.9.1 are not changed anything of callback's work, just is added the new calls of callback with internal instance for shutdown gracefully based on the 8.7.1 work.

so,  if I only need to do the same work as 8.7.1 with the code of 8.9.1, I can just ignore to use the CURLOPT_PRIVATE if curl instance is the internal instance, for exampe:
The old code:
...
curl_easy_getinfo(curl, CURLINFO_PRIVATE, (void *)&request);
void *cert = request->cert; //No check the output parameter.

I can just change it to:
...
curl_easy_getinfo(curl, CURLINFO_PRIVATE, (void *)&request);
if (request) // Just add the conditional checking statement to ensure the request isn't null before use it.
{
...
}

Could you please let me know is that the right idea?

Thanks very much.

**Comment by icing** [comment]
This looks like the right idea to handle this. If you set CURLINFO_PRIVATE on your requests, you will see the "internal" handles as having `NULL` for this value.
